### PR TITLE
KEYCLOAK-14665 initial spi interfaces with validator chain and UserAttributesReprese…

### PR DESCRIPTION
**Pull request by bs-matil**
This is the initial contribution for the Milestone M1 for the legacy profile as documented in https://github.com/keycloak/keycloak-community/blob/master/design/user-profile.md and relates to https://issues.jboss.org/browse/KEYCLOAK-14665

A specific Ticket will be created and added soon.

Contents:
* SPI which supports only methods needed to support the legacy profile e.g. the available attributes feature is not implemented yet as nothing would evaluate it. 
* Combination of all validation logic into one SPI implementation
* SPI with ValidationScenario handling for different places of validation (e.g Registration, IdpReview etc.) 
* A ValidaitorChainBuilder to hardcode the validation rules as long there is no configuration mechanism in place

### CODE REVIEW

**Comment by abstractj** [comment]
@bs-matil is this PR covering the whole epic here https://issues.jboss.org/browse/KEYCLOAK-2966 ? Or part of this? For now I'm adding the label missing, Jira, but please let me know if my assumption is completely wrong.

**Comment by stianst** [comment]
@abstractj It's just doing first part of it. We should have a JIRA to describe what it is and associate this PR/commit with it.

**Comment by stianst** [comment]
@bs-matil I ran out of time before hols' so won't be able to take a look at least before July I'm afraid.

**Comment by bs-matil** [comment]
I just refactored all against the new attribute handling in the user model

File: services/src/main/java/org/keycloak/userprofile/LegacyUserProfileProvider.java
Comment (bs-matil) [inline services/src/main/java/org/keycloak/userprofile/LegacyUserProfileProvider.java:80]: This is needed because of the "chained action" approach in the registration. May it should be considered to adapt the registration flow in order to remove this switch so we can go here with configuration only.

**Comment by bs-matil** [comment]
@stianst any news on regards that pr?

**Comment by stianst** [comment]
@bs-matil No, sorry, we haven't gotten around to reviewing this yet I'm afraid. I aim to take a look at this tomorrow (or Monday at the latest). @mposolda have you had a chance to look at this yet?

**Comment by mposolda** [comment]
@stianst @bs-matil Sorry, I did not yet look at this. I hope to take a look once Wildfly 20 upgrade is finished (should be early next week hopefully).

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java
Comment (stianst) [inline services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java:83]: I think looking for specific error messages like this is too brittle, perhaps it would be better to have a `UserProfileValidationException` with sub-classes like `EmailInUseException` instead here?

File: services/src/main/java/org/keycloak/userprofile/validation/Validator.java
Comment (bs-matil) [reply services/src/main/java/org/keycloak/userprofile/validation/Validator.java:27]: Yes indeed. We have this locally for the "new" profile. We wouldn't introduce it with this PR to minimize the changes.

File: server-spi-private/src/main/java/org/keycloak/userprofile/validation/UserProfileValidationResult.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/validation/UserProfileValidationResult.java:26]: validate also returns state changes e.g. an attribute has been changed. This is needed for decision making later on. Alternatively we would have validation results and change results grouped in a general result class.

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java:34]: What would you prefer instead? A UserUpdateProvider? To me, it was important to have a unique entry point for updating users as the old behavior was very error-prone (and yes I found "relevant" errors in the different code paths which are gone now). 
I would prefer a single stable update method. Rather in this provider or its own one. This also would enable profile implementations to execute actions on save e.g. external validation or data source checks. But may I'm wrong here.

File: server-spi-private/src/main/java/org/keycloak/userprofile/AbstractUserProfileProvider.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AbstractUserProfileProvider.java:27]: It will have one since we introduce the new profile.. but sure we can leave it out till then. I will remove it.

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileSpi.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileSpi.java:46]: Okay, we were not sure about that (both patterns are applied in the codebase and its hard to guess when it's appropriate).

will be fixed.

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java:32]: I will do a refactoring of this solution.
But I would prefer to introduce no"empty" value UserProfiles for getCurrent() as it would be quite complex to give the profile the ability to estimate how to create a user in the given context. Therefore I would expect that the caller always initialized the new user before calling the validation function.

File: services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java
Comment (bs-matil) [reply services/src/main/java/org/keycloak/authentication/forms/RegistrationProfile.java:83]: I would count that into the category of improvements we could do also later as the given behavior is quite near to what was there before. But I can also put it in this PR.

File: server-spi-private/src/main/java/org/keycloak/userprofile/AbstractUserProfileProvider.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/AbstractUserProfileProvider.java]: done

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java]: @stianst please have a look if the current interface is more as expected.

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java
Comment (stianst) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProvider.java]: I agree it would be nice to have a single place that deals with updating users, we probably don't need an SPI for it. As simple way for now could be to just add a UserUpdateHelper or something, that can be refactored later as we are planning to clean up the stores quite a bit, especially around this sort of stuff. @hmlnarik wdyt?

**Comment by Martin-Idel-SI** [comment]
(I also posted the following to the user mail discussion thread):

Our rationale with the PR is the following:
-	The PR essentially implements the first milestone in the design (except for a “final” implementation of the SPI interface, see https://github.com/keycloak/keycloak-community/blob/master/design/user-profile.md)
-	We intentionally made the SPI private to allow changing it with future PRs. We feel that the feature is too big to confidently and irrevocably design the interfaces right now. If having it as a private SPI is not enough, we could also mark all interface classes/methods as “Deprecated” to further inhibit their use for now.
-	After Milestones M2 and M3 are implemented, the interfaces should have emerged and we could pin the exact API and put the SPI in public. Any further changes should then be additions. We believe that this would make progress quicker. 
-	This is true especially for the validation (this is milestone M2). The current “validation” SPI was only created to gather all current validation into the LegacyUserProvider and I believe it works well for that purpose. I realise that the title of the PR is somewhat unfortunate in that respect.
-	We should ensure that this PR works together with Thomas Darimont’s PR for validation (https://github.com/keycloak/keycloak/pull/7324) when it gets merged. However, we believe that this is independent: We could merge this PR and the other PR and then integrate the validation into the user profile without any problems in another PR if the fundamental design of the validation works with the validation logic which is currently on master. Hence the question whether the validation logic works with this PR can be decided independently and we would like to move forward with the User Profile PR.
What do you think? What is still missing in your eyes or do you have a problem with our approach?

**Comment by pedroigor** [comment]
> @pedroigor any news on that?
> 
> Should I may refile the PR, the history is a bit cluttered.

@bs-matil I'll be looking User Profile this week, starting today. Need to sync what was discussed/done so far before reviewing this PR or any other that is related.

One thing though, from your comments it seems this PR is delivering what was defined as M1 in the design document. I agree that we can deal with validation later (including the PR from @thomasdarimont around validation) and focus on the core changes to support user profiles.

**Comment by pedroigor** [comment]
@bs-matil, There are a lot of changes to formatting/cleanup and empty lines being removed that do not seem related to the changes you are doing.

In order to make the reviewing process a bit easier and objective, could you please revert any change like that?

**Comment by bs-matil** [comment]
> 
> 
> @bs-matil, There are a lot of changes to formatting/cleanup and empty lines being removed that do not seem related to the changes you are doing.
> 
> In order to make the reviewing process a bit easier and objective, could you please revert any change like that?

I'll work on this.

**Comment by bs-matil** [comment]
> 
> 
> > @pedroigor any news on that?
> > Should I may refile the PR, the history is a bit cluttered.
> 
> @bs-matil I'll be looking User Profile this week, starting today. Need to sync what was discussed/done so far before reviewing this PR or any other that is related.
> 
> One thing though, from your comments it seems this PR is delivering what was defined as M1 in the design document. I agree that we can deal with validation later (including the PR from @thomasdarimont around validation) and focus on the core changes to support user profiles. Does it make sense for you?
> 
> Btw, could you please rebase?

rebase has been completed. tests pending.

File: services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java:130]: I think `UserProfileUpdateHelper` can be simpler if we just pass the attribute map instead of the `UserModel`. It would avoid having to create this anonymous class to work around the contract to only fetch attributes.

I'm also wondering why the logic on `UserProfileUpdateHelper` is not on the `UserProfileProvider` itself?

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProviderFactory.java
Comment (pedroigor) [inline server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProviderFactory.java:25]: I think we could have some methods here to create the `UserProfileProvider` instance without pushing too much to clients of the API. For instance:

```
Map<String, List<String>> newAttributes = // obtain attributes to update
UserProfileProviderFactory factory = // obtain factory from session
UserProfileProvider provider = factory.create(UserUpdateEvent.IdpReview, user.getAttributes(), newAttributes);
UserProfileValidationResult result = profileProvider.validate();
```

This can hide some implementation details such as the `UserProfileContext` implementation that should be used.

In addition to that, the user ID also works like any other attribute into the attribute map that is the real source of data and data structure we care about to perform the validations. 

The `Attributes` API I gave as an example could also help to make this SPI more attribute oriented and simplify even more how clients pass over the attributes from the different places in Keycloak as well as using different sources.

**Review by pedroigor — CHANGES_REQUESTED** [review]

**Comment by pedroigor** [comment]
@bs-matil These are my initial thoughts, I hope they make sense for you. I'll hold a bit the review and wait for your feedback considering that some of my comments are related to the design of the SPI.

File: services/src/main/java/org/keycloak/userprofile/utils/StoredUserProfile.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/userprofile/utils/StoredUserProfile.java:30]: I understand the reason behind it I think. So that a `UserProfile` is marked as writable.

After reading through your code, looks like to me that the best would be if we had a single `UserProfile` implementation, with attributes encapsulated via a `Attributes` API so that the read/write capability could be controlled accordingly depending on the type of the event.

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java:59]: Why create an attribute with a list of `null` values ?

**Review by pedroigor — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/keycloak/representations/account/UserRepresentation.java
Comment (bs-matil) [reply core/src/main/java/org/keycloak/representations/account/UserRepresentation.java:102]: the check does not check for attribute values but for the attribute list. If the list is not initialized (which is likely due to the nature of the Representation objects) this should IMHO nothing the user of this function has to deal with.

File: core/src/main/java/org/keycloak/representations/idm/UserRepresentation.java
Comment (bs-matil) [reply core/src/main/java/org/keycloak/representations/idm/UserRepresentation.java]: same here its not a value check

File: services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java
Comment (bs-matil) [reply services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java:59]: because for example this test expects that the handling of email is "" -> null while this is not the case anywhere else... 
https://github.com/keycloak/keycloak/blob/74988a3f21c26764e01956ff8c9ef5cea865cb2d/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/forms/ResetPasswordTest.java#L836

Comment (bs-matil) [reply services/src/main/java/org/keycloak/userprofile/utils/UserProfileUpdateHelper.java:59]: I invested time in keeping the api of the validation as consistent as possible. But there are inconsistencies in the code 

Thats why I initially put the user update method into the user profile. That would have enabled to override the behaviour for the old very inconsistent code path while having a clean new store method in the future. 

But Stian argued against this as the profile should not care about storage but validation only. So I moved the code to this helper class and out of the SPI interface.

File: services/src/main/java/org/keycloak/userprofile/utils/StoredUserProfile.java
Comment (bs-matil) [reply services/src/main/java/org/keycloak/userprofile/utils/StoredUserProfile.java:30]: I'm a bit confused to be honest. the current UserProfile interface is actually very close to an Attributes interface. So to move towards your solution I would rename UserProfile to UserAttributes and merge the StoredUserProfile methods into this interface. Then on all implementations I would add the write methods and throw Exceptions when they are called.

Is this even close to your idea? I will do so. But may I got you fully wrong here, I at least suspect that.

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfile.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfile.java:29]: As I tried to explain in the other comment regards the interface of the userprofile. The current UserProfile interface is actually just an attributes interfaces (may a bad one) but I'm right now not sure how to go towards your solution.  I will push what I understood soon..

File: services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java
Comment (bs-matil) [reply services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java:130]: Regards the position of user profile itself see other comment.

Regards the AttributeList.
This won't work because of the nature of the different UserModel implementations (e,g, for idpReview). Some just are not useable as simple attribute list. (Their interface for the attributes does not work consistently)

Comment (bs-matil) [reply services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpReviewProfileAuthenticator.java:130]: The differences are reflected in the different user profile implementations

File: server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProviderFactory.java
Comment (bs-matil) [reply server-spi-private/src/main/java/org/keycloak/userprofile/UserProfileProviderFactory.java:25]: @pedroigor  I tried to go into this direction. It's not possible to do this on the SPI level as the interfaces would have to know the concrete types in the service impl. So I moved the factory like methods into the DefaultUserProfileContext

**Comment by bs-matil** [comment]
> 
> 
> @bs-matil These are my initial thoughts, I hope they make sense for you. I'll hold a bit the review and wait for your feedback considering that some of my comments are related to the design of the SPI.

I would wait now for additional feedback @pedroigor

**Review by pedroigor — APPROVED** [review]
LGTM. Thanks, @bs-matil.

I think it is a very good start and unblock us to start working on the next round(s) of changes.

File: services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:233]: @bs-matil I forgot to mention this change in my last review. Why did you add this endpoint for?

Comment (bs-matil) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:233]: looks like accidentally. Shall I open a PR for removal?

Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:233]: Yes, please. Sorry about that. I added the comment but it got lost during the review ...

Comment (bs-matil) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:233]: Sorry too, I tought I will at least have one more time to squash and cleanup. Wasn't expecting a so fast merge after the history of this PR

Comment (bs-matil) [reply services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java:233]: https://github.com/keycloak/keycloak/pull/7467
PR for cleanup

**Pull request merged**
